### PR TITLE
Reuse the residual linearization across Jacobian columns; separable field-line synthesis (S1 piece 2)

### PR DESCRIPTION
**Pull request by rogeriojorge**
S1 piece (2) of the 2026-09-13 plan: re-lands #299's residual-linearization reuse and separable Boozer field-line synthesis on `main`, without #299's plan, logbook, record or benchmark-harness edits.

## Problem

- `_implicit_evolved_tangent_multi_rhs` (the block/GMRES Jacobian column solve) called `jax.jvp` of the preconditioned residual afresh for every column's right-hand side and for every Krylov matvec. Each call re-linearized the full residual, including the preconditioner, at the same `(z*, p)`.
- Four callers synthesized `|B|` along Boozer field lines by building an `(s, α, φ, mode)` phase tensor, then `cos`/`sin` and an `einsum`: `optimize._qi_grid`, `omnigenity.omnigenity_residual`, `bounce.trace_boozer_field_lines` and `maxj.common_trapped_pitches_state`. Under reverse mode that 4-D tape is the largest allocation of the QI residual Jacobian.

## Change

- `vmex/core/implicit.py`: `jax.linearize(residual, z_star, params)` once, then reuse the linear map for the right-hand side (`residual_tangent(0, dp)`) and the matvec (`residual_tangent(dz, 0)`). The derivative of the preconditioner itself (`dP·F` at the finite residual) is retained; the preconditioner is not frozen.
- `vmex/core/bounce.py`: `_boozer_field_strength(bmnc, bmns, xm, xn, iota, α, φ)` uses `mθ − nφ = mα + (mι − n)φ`, which separates α from the surface-dependent phase. Both parities, physical `n` and the ι derivative are kept, and `einsum` uses highest precision for float32. The four callers above now use it; `_qi_grid` swaps axes to keep its `(s, φ, α)` layout.
- `tests/test_bounce.py`: `test_field_line_synthesis_matches_dense_values_and_derivatives` compares the separated synthesis with the dense reference, both parities × nfp 1 and 3 × float32 and float64 (value, JVP and VJP under `jax.jit`, plus a float64 central difference). Tolerances: 2e-11 for float64; for float32, 1e-4, with the VJP compared as a norm at 2e-5.

## Evidence (from #299, not re-measured here)

Source commit `ef8a25fa`. #299's PR body credits "full residual linearization reuse" to `f205c995`, but that hunk is in `ef8a25fa`; `f205c995` is S1 piece 4. Record entries `changes`, `component_48_dof`, `gpu_trace` and `gpu_fieldline_kernel` in [`benchmarks/review_optimization_20260912.json`](https://github.com/uwplasma/vmex/blob/perf/reuse-objective-linearization/benchmarks/review_optimization_20260912.json) on #299's branch:

- Warm 48-dof QI Jacobian on CPU (nfp = 1, ns = 16, mpol = ntor = 5, QI surrogate on four surfaces, refinement disabled, six samples) goes from a 1.098 s median to 0.818 s. Peak RSS goes 5,438 → 3,702 MiB. Last-sample parity: residual 1.1e-13 and Jacobian 5.7e-14 relative L2.
- Field-line kernel alone on GPU (`gpu_fieldline_kernel`: shape 4 × 27 × 141 × 648, float64):
  - VJP: 5.05 → 0.81 ms median, 159 → 17.5 MB temporaries, compile 1.09 → 0.49 s;
  - value: 3.84 → 0.50 ms, temporaries 0.97 → 9.2 MB.
- There, #299 validated the GPU dense-reference value/JVP/VJP tests (float32 and float64, both parities, nfp 1 and 3), six QI and maximum-J physics AD tests, and the implicit block, transpose and chunk checks.

Those are #299's component rows from a private harness. The A1 counter rows below were measured on this branch after #310 merged.

### A1 counter rows (`benchmarks/optimization.py`, #310)

Command, run from each checkout in fresh processes, alternating main / #318 / main / #318 per case:

```bash
OMP_NUM_THREADS=4 OPENBLAS_NUM_THREADS=4 XLA_FLAGS=--xla_cpu_parallel_codegen_split_count=128 \
VMEX_COMPILATION_CACHE=disabled PYTHONPATH=. \
python benchmarks/optimization.py --case {qa,qi} --max-mode 1 --optimizer none --nfev 2
```

Environment: Apple-silicon laptop, Python 3.12.9, JAX 0.11.1, NumPy 2.5.3, SciPy 1.18.1, vmex 0.8.1; the thread settings of #310's committed record. Baseline `main` is `08d92161`, candidate `1e0eea21` (this branch merged with `08d92161`); both checkouts were clean. Load averages ran 5.7–18.7 from other sessions, so **seconds are diagnostic**; counts and identities are the gate. 8 dof; each cell gives repeat 1 / repeat 2.

| QA | main `08d92161` | #318 `1e0eea21` |
|---|---|---|
| build: 1 solve, 185 iterations | 5.31 / 5.60 s, 244 compiles | 5.30 / 5.68 s, 244 compiles |
| first derivative, wall | 46.57 / 52.53 s | 44.30 / 50.09 s |
| · refinement (3 steps, 6,000 Krylov iterations) | 22.82 / 26.93 s | 22.61 / 25.15 s |
| · Jacobian (1 call, 8 columns, 0 certifier iterations) | 23.44 / 25.24 s | 21.40 / 24.63 s |
| · compiles, compile seconds | 33; 34.35 / 37.53 s | 33; 32.00 / 36.68 s |
| JAX value-and-gradient lane | 37.71 / 41.00 s (2 compiles) | 36.90 / 41.19 s (2 compiles) |
| whole run: compiles, compile seconds, wall | 281; 75.35 / 82.23 s; 90.96 / 100.68 s | 281; 72.40 / 80.92 s; 87.92 / 98.38 s |

| QI | main `08d92161` | #318 `1e0eea21` |
|---|---|---|
| build: 1 solve, 185 iterations | 11.16 / 9.77 s, 500 compiles | 10.24 / 9.89 s, 504 compiles |
| first derivative, wall | 56.46 / 50.31 s | 51.87 / 50.41 s |
| · refinement (3 steps, 6,000 Krylov iterations) | 25.56 / 22.63 s | 24.75 / 23.28 s |
| · Jacobian (1 call, 8 columns, 0 certifier iterations) | 29.77 / 26.70 s | 26.08 / 26.12 s |
| · compiles, compile seconds | 39; 41.96 / 37.98 s | 39; 37.26 / 38.23 s |
| JAX value-and-gradient lane | 49.61 / 46.70 s (2 compiles) | 45.87 / 46.09 s (2 compiles) |
| whole run: compiles, compile seconds, wall | 543; 99.47 / 91.51 s; 118.90 / 108.34 s | 547; 90.55 / 91.44 s; 109.47 / 107.90 s |

- **Unchanged:** solves (1, 185 iterations), refinement steps and Krylov iterations, Jacobian calls and columns, and adjoints (0) are identical in every run. The JAX lane reports no counters, since A1 does not count inside compiled programs.
- **Identity at x0:** the harness records the host value and gradient norm, not the gradient array.
  - QA: value `0.6268904087727508` and gradient norm `5.590597528997621` are bit-identical between main and #318.
  - QI: value `1.204542660467874` vs `1.2045426604678768` (2.4e-15 relative), gradient norm `277.6471729965654` vs `277.64717299656826` (1.0e-14 relative). That is the rounding change of the separated field-line synthesis.
  - Repeats within each branch are bit-identical, and the JAX lane and differentiated graph match the host to ≤ 1.8e-16 on both branches.
- **Compiles:** QI builds compile 4 more programs (500 → 504); first-derivative and JAX-lane compile counts are unchanged.
- **Reading:** at 8 dof the first derivative is cold and compile-dominated (32–42 s of 44–56 s), and the Jacobian differences sit inside run-to-run spread at this load. These rows do not test #299's warm 48-dof QI Jacobian claim (1.10 → 0.82 s): the harness has no warm evaluation at `--optimizer none`. Treat that claim as #299's component evidence only.

### Warm 48-dof QI measurement (merge-rule run)

The merge rule for this PR: the warm QI Jacobian improves beyond run-to-run spread, OR peak RSS drops materially, with identical counts.

**Setup.**
- Command: `benchmarks/optimization.py --case qi --max-mode 3 --optimizer least_squares --nfev 3`. That is 48 dof, with three warm Jacobians inside the optimizer after the first derivative.
- Runs: fresh processes, alternating main / #318 / main / #318.
- Code: main `2fbe4e57` (with #328 and #320) against this branch merged onto it, `401ab149`. Both checkouts were clean.
- Machine: the 36-core Linux workstation, with Python 3.12.13 / JAX 0.11.1 / NumPy 2.5.2 / SciPy 1.18.1.
- Settings: `JAX_PLATFORMS=cpu`, 4 threads, `XLA_FLAGS=--xla_cpu_parallel_codegen_split_count=128`, and `VMEX_COMPILATION_CACHE=disabled`.
- Peak RSS comes from `/usr/bin/time -v`.
- An external wrapper, identical for both arms, records every `residual` and `residual_jac` call's wall time and A1 counter delta. The harness itself is unchanged.
- A run counts only if the 1-minute load is ≤ 24 at both its start and its end. All four runs qualify (loads 11.9–23.8).

**Discarded attempts.** A laptop attempt at load 24–140 was discarded under the load rule. A first attempt at main `9a6b5efc` crashed on both arms in the JAX contract phase with a 1.52 TB request. That was the reverse-Jacobian fallback bug fixed by #328.

| run | load start → end | peak RSS | first derivative (Jacobian part) | warm Jacobians 1 / 2 / 3 | optimize phase |
|---|---|---|---|---|---|
| main rep 1 | 11.9 → 14.0 | 4.18 GiB | 104.2 s (69.6 s) | 5.37 / 4.59 / 4.54 s | 64.4 s |
| #318 rep 1 | 14.0 → 12.9 | 4.02 GiB | 88.5 s (58.5 s) | 3.81 / 4.93 / 4.13 s | 68.9 s |
| main rep 2 | 12.9 → 21.9 | 4.19 GiB | 101.5 s (70.2 s) | 5.08 / 5.04 / 5.17 s | 69.0 s |
| #318 rep 2 | 21.9 → 23.8 | 3.96 GiB | 112.8 s (75.8 s) | 5.14 / 5.01 / 5.25 s | 73.0 s |

- **Warm Jacobian:** main's run means are 4.83 and 5.10 s (pooled 4.97 s). #318's are 4.29 and 5.13 s (pooled 4.71 s, −5 %). The arms overlap: #318's second repeat is slower than main's first. **This does not improve beyond run-to-run spread.**
- **Peak RSS:** 4.18 / 4.19 GiB on main against 4.02 / 3.96 GiB on #318. That is 0.16–0.23 GiB (4–5.5 %) lower and consistent across repeats. It is well short of the 5,438 → 3,702 MiB (−32 %) that #299 recorded on its 48-dof collaborator case.
- **Counts:** identical in every run and on both arms, except refinement Krylov iterations. Those are 9,444 on main against 9,418 on #318 (−26), consistent within each arm.
  - The same in every run: 3 solves / 3,852 iterations; 4 Jacobians / 192 columns / 0 certifier iterations; 3 refinements / 6 steps; first-derivative, optimize and JAX-lane compiles of 28, 10 and 2.
  - Build compiles differ: 508 against 512.
- **Identity:** initial cost 1.1e-15 relative, gradient norm 8.5e-16, final cost after 3 evaluations 4.0e-10. Each arm's two repeats are bit-identical.
- **Verdict under the rule:** not demonstrated. The warm Jacobian gain is inside run-to-run spread. The peak-RSS drop is real but about 5 %, which I do not count as material. The counts differ by 26 refinement Krylov iterations, from the synthesis rounding.

Local verification (this branch, Apple silicon CPU, Python 3.11, JAX 0.9.2, shared machine under load 15–40):
- `tests/test_bounce.py` gives 21 passed in 53 s (the 8 new parametrizations included).
- A focused selection of the changed callers gives 17 passed and 1 skipped in 220 s. It covers `tests/test_qi_maxj_composition.py`, `tests/test_qi_reference_oracle.py`, three `test_maxj.py` tests (common trapped pitches ×2, jit and AD vs FD), three `test_omnigenity.py` tests (analytic QI zero, state gradient, lasym QI derivative) and `test_optimize.py::test_least_squares_implicit_jac_chunking`. The lasym omnigenity test is marked `full` and was skipped.
- After merging `main` at `08d92161` (which brings #310's counters, #314 and #323), the selection above plus `test_optimize.py::test_least_squares_implicit_jac_solver_block` (block and GMRES Jacobians through the changed tangent solve) and #323's `test_qi_regression_pin_and_jit` (which reads the changed `_qi_grid`) gives 40 passed in 360 s. Python 3.12 / JAX 0.11.1: `tests/test_bounce.py` plus the QI pin give 21 passed, 1 skipped (the DESC comparison test, `importorskip("desc")`, since DESC is not installed there); the QI pin itself passes there.
- `python tools/preflight.py --static` passes (80 passed, 3 skipped); `python tools/test_manifest.py check` passes.

## Limitations

- Warm Jacobian only. #299 measured no cold or first-derivative change, and no whole-optimization speedup is claimed.
- The field-line kernel's *forward-only* temporaries rise (0.97 → 9.2 MB on the GPU kernel below) while its derivative temporaries fall. Do not confuse this change with the separable *geometry* Fourier synthesis in the row Jacobian, which #299 measured slower on CPU and GPU and did not land (record `warm_gpu_source_attribution`).
- `jax.linearize` keeps the residual's primal intermediates alive for the whole column solve, trading memory held for recomputation. Peak RSS fell on the recorded case, but it was not measured on free-boundary or high-resolution decks.

## What to check

- `residual_tangent(zero_state, tangent)` and `residual_tangent(value, zero_params)` equal the two removed `jax.jvp` calls. The linear map is taken jointly in `(z, p)`, so the `p` part stays exactly what it was.
- `_boozer_field_strength` with `bmns=None` matches the old callers that passed zeros (`maxj`) or a sine table (`bounce`, `omnigenity`, `_qi_grid`).
- `_qi_grid`'s `swapaxes(…, 1, 2)` restores its `(s, φ, α)` layout.

### CODE REVIEW

**Comment by rogeriojorge** [comment]
Closing under the merge rule recorded in plan.md: merge only if the warm Jacobian time or peak memory improves beyond run-to-run spread.

Office A/B/A/B on the 48-dof QI case (main 2fbe4e57 vs 401ab149, JAX 0.11.1, 4 threads, cache disabled, every run at 1-minute load ≤ 24): warm Jacobian run means were 4.83 / 5.10 s on main and 4.29 / 5.13 s here, inside spread. Peak RSS fell 4–5.5 % (4.18–4.19 to 3.96–4.02 GiB), well short of the 32 % in #299's record and not enough for +90/−31 lines. Counts matched except refinement Krylov iterations (9,444 vs 9,418). The table is in the PR body; the source stays in this branch's history.
